### ISSUE TITLE
TTA can improve the accuracy of Image classification

### ISSUE THREAD

**Issue by canhnp** [andrewekhalel/edafa#6]
Dear,

I have tried TTA
I seems that it achieves the low accuracy with TTA
How do you think?

Best regards

**Comment by canhnp** [comment]
Dear Andrew,

For training, I use this one:

 imgaug.Affine(shear=10.0,
                #rotate_max_deg=45,
                rotate_max_deg=179,
                interp=cv2.INTER_CUBIC,
                border=cv2.BORDER_REFLECT),
            imgaug.Flip(horiz=True),
            imgaug.Flip(vert=True),
            imgaug.ToUint8(),


For test prediction, I apply rotate 90,180, 270 only.

# if can fit normally generate batch

    # x is an array of images i.e img0, img1, img2 etc.

    batch_id = 0

    batch_size = 16

    y_prob_list = []

    for degree in [0, 45, 90, 135, 270, 315]:

        if degree == 0:

            x_tta = x

        else:

            x_tta = [rotate(x[i]) for i in range(0, x.shape[0])]

        y_prob = np.empty(y_true.shape)

        for i in range(0, x.shape[0], batch_size):

            x_batch = x_tta[i:i+batch_size]

            y_prob[i:i+batch_size] = predictor([x_batch])[0]

        x_batch = x_tta[i:] # whatever remain

        y_prob[i:] = predictor([x_batch])[0]

        y_prob_list.append(y_prob)

    # ensemble, essentially averaging probability maps

    y_prob_list = np.array(y_prob_list) # shape should be (N_tta,
N_samples) + sample prediction dims

    y_pred = np.mean(y_prob_list, axis=0) # average across N_tta per
sample, can do max/voting etc. instead# if can fit normally generate batch

    # x is an array of images i.e img0, img1, img2 etc.

    batch_id = 0

    batch_size = 16

    y_prob_list = []

    for degree in [ 90, 180, 270]:

        if degree == 0:

            x_tta = x

        else:

            x_tta = [rotate(x[i]) for i in range(0, x.shape[0])]

        y_prob = np.empty(y_true.shape)

        for i in range(0, x.shape[0], batch_size):

            x_batch = x_tta[i:i+batch_size]

            y_prob[i:i+batch_size] = predictor([x_batch])[0]

        x_batch = x_tta[i:] # whatever remain

        y_prob[i:] = predictor([x_batch])[0]

        y_prob_list.append(y_prob)

    # ensemble, essentially averaging probability maps

    y_prob_list = np.array(y_prob_list) # shape should be (N_tta,
N_samples) + sample prediction dims

    y_pred = np.mean(y_prob_list, axis=0) # average across N_tta per
sample, can do max/voting etc. instead


Best regards,
Canh Nguyen

Vào Th 7, 29 thg 12, 2018 vào lúc 16:24 Andrew Khalel <
notifications@github.com> đã viết:

**Comment by andrewekhalel** [comment]
Great results! however, there is no guarantee that all transformations will improve results. So you should choose the set that maximize the accuracy (3 in this case). It looks like horizontal flips are not effective in your scenario.

Best,
Andrew

**Comment by andrewekhalel** [comment]
For me it looks okay. However, it would be better if you apply flipping on validation set to see whether it enhances results or not before applying to test set.

Andrew

**Comment by canhnp** [comment]
Dear Andrew,

Thanks for your info.

But, I'm not clear about average of prediction in your method. In my case, trained model will predict the class probability (e.g. image classification with 4 class) -> So, I think that we should use voting, maximum of probabilities or sum of probabilities to predict final label of test image

What do you think?

Best regards,
Canh Nguyen

**Comment by Linamiao-1998** [comment]
Sorry to excuse me , could you tell where should I use TTA , in the model ?  in the image ? or other place ? and how should I use TTA ,please.
